# Top selling cars in germany by brand 
In the following project we will analyze a list of cars from Ebay in germany, we will initially clean the data and then understand the success of the top selling brands 

In [41]:
import pandas as pd
import numpy as np

We want to use the pandas module to open the document "autos.csv". Note that if we simply execute the pd.read_csv command, we will have an error;
That error happens  because Python can only read binary code in its lowest levels and there is non-binary code in the data frame, therefore we have to use the argument "encoding"  to avoid this problem. Remember that the 3 most common types of encodings are 
* "UTF-8"
* "Latin-1"  also know as (ISO-8859-1)
* "Windows 1251"

After trying to use "UTF-8" (which did not work) , we used "Latin-1" and that one worked. In the following code we will view the first five columns using the `Dataframe.head(5)` and the `Dataframe.info()` command to have a better understanding of the data frame we will manipulate. 

In [42]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The most obvious information that we immediately see is that our data frame has 5000 rows and 20 columns. Also notice that most of the entries' types are *object*, that basically means that there is an string in that entry. And, when it says *int64* that is just an special type of designation used in numpy for *integer* but is not anything else than that, just an integer. 


# Fixing names of the heading 

We will see first what names we have in the heading row. 

In [43]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We will change the camel case to snake case and we will also change some names. 

In [44]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Basic Data exploration 

In [45]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

At a first glance of the data, notice that 
* *seller*  and *offer_type* column seem to have the same value. 
* Also notice that *num_photos* is not very informative since it seems to have a 0 in every column

In [46]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

Therefore we will get rid of *seller*, *offer_type*  and *num_photos* since those columns lack of relevant information. 

In [47]:
autos=autos.drop(["seller","offer_type","num_photos"], axis = 1)

There are two columns (`odometer`  and `price` ) whose values are numeric but have strings which prevent us from analyzing the data therefore we will remove those extra characters in order to manipulate the data easily. 

In [48]:
autos["odometer"].head(3)

0    150,000km
1    150,000km
2     70,000km
Name: odometer, dtype: object

In [49]:
autos["price"].head(3)

0    $5,000
1    $8,500
2    $8,990
Name: price, dtype: object

Afer viewing the entries from the afore mentioned columns we conclude that the special characters *km* and "," will be removed from the `odometer` column, and the special chatacters *$* and "," will be removed from the `price` column. We will also rename the `odometer` column to `odometer_km`, to know the units of the column. 

In [50]:
autos["odometer"]=(autos["odometer"]
                         .str.replace("km","")
                         .str.replace(",","")
                         .astype(int)
                  )




In [51]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace = True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Now, notice that with the column `price`, which is a numeric price we have some symbols that will prevent us from doing our analysis as well. 

Now, notice that with the column `price`, which is a numeric price we have some symbols that will prevent us from doing our analysis as well. 

In [52]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

The elements that we need to get rid of are the "$" symbol and the ",". 

In [53]:
autos["price"]=( autos["price"]
                .str.replace("$","")
                .str.replace(",","")
                .astype(int)
               )

Now, lets see how that looks. 

In [54]:
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

# Exploring odometer and price

In [55]:
print(autos["price"].unique().shape)
print(autos["odometer_km"].unique().shape)

(2357,)
(13,)


Initially we can see from that information that there are 2357 different prices and just 13 different odometers therefore we will need to use different commands to have a more global understanding of the information contained. 

In [56]:
print(autos["price"].describe())
print(autos["price"].value_counts().head(5))

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64


In [57]:
autos["odometer_km"].value_counts().head(10)                                        

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
Name: odometer_km, dtype: int64

It is easy to see that the maximum and minimum of the `price` column are not correct, it does not make sense that the price of a car is "0" actually there are 1421 cars whose price is "0" given that in total there are 50000 cars that is about 3% therefore we should consider removing those rows, and it does not make sense either that the price of a car is around a million. Also, notice that all values are rounded, which means that customers may have selected the prices from a given list of prices. 

Now, about "odometer_km" notice that all values are rounded as well, which indicates that customers may have chosen the values from a preselected list as well. Notice that there are more high mileage vehicles than low mileage vehicles.  

Based on that information we will sort both lists to have a better understanding o the frequency of top and min values. 

In [58]:
print(autos["price"].value_counts().sort_index(ascending=False).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


In [59]:
print(autos["price"].value_counts().sort_index(ascending = True).head(20))


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


In [60]:
print(autos["price"].value_counts().sort_index(ascending = False)[0:13])
print("There about",
      autos["price"].value_counts()
      .sort_index(ascending = False)[0:10].sum(),
      "with values around a million")

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
Name: price, dtype: int64
There about 14 with values around a million


Since eBay is an auction site, there could be cars whose opening bid is a million, but we will remove anything above 350.000 to manipulate more realistic values. Similarly we will also remove cars whose price is 0. 

In [61]:
autos=autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the date columns 

In this data set there are 5 columns devoted to dates, "date_crawled","ad_created","last_seen", "registration month"  and "registration_year", registration columns are numeric and the other columns are strings. We want to explore those columns to understand the date ranges. 

In [62]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Before exploring for example "date_crawled", notice that we are interested in the date that it was crawled, the hour does not play any role for our analysis, that in formation is contained into the 10 first characters of our string, that is the reason because we will use the `str` accessor, notice on the next examples how its use can affect the final result. 

In [63]:
autos["date_crawled"].head(3)

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
Name: date_crawled, dtype: object

In [64]:
autos["date_crawled"].str[:5].head(3)

0    2016-
1    2016-
2    2016-
Name: date_crawled, dtype: object

In [65]:
autos["date_crawled"].str[:10].head(3)

0    2016-03-26
1    2016-04-04
2    2016-03-26
Name: date_crawled, dtype: object

Therefore we will use the `str` accessor to print the date only. Notice as well that when we use the `value_counts()` function we will not print the frequency, we will rather include the percentage. 

In [66]:
(autos["date_crawled"].str[:10]
                     .value_counts(normalize = True ,  dropna = False)
                     .sort_index())        

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [67]:
(autos["date_crawled"].str[:10]
                      .value_counts(normalize = True, dropna = False )
                      .sort_values()) 

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

From the two codes just executed when can see that the data was crawled consistently from march the 5th until april the seventh and also note that the frequency roughly ranged from 0.0014 until 0.038. We will now analyze the same information for the "last_seen" and "ad_created"columns. 

In [68]:
(autos["last_seen"].str[:10]
                   .value_counts(normalize = True, dropna=False)
                   .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The "last_seen" column indicates the last time crawled by the crawler, which allow us to know when a listing was removed, presumabl because the car was sold. The last three days are disproportionate to the other days, which possibly indicate a flaw on the systems rather than a massive sold. 

In [69]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Note that most "ad_created" dates fall within one-two months of the listing date, with some exceptions being the oldest at around 9 months. 

In [70]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year the the car was first registered will indicate the age of the car. Note as well that there are some odd values in that list, for example the min is 1000 , way before the car was invented, and the maximum would be 9999, many years into the future. 

# Dealing with Incorrect Registration Year Data. 

Because a car can not be first registered before the listing was seen, any vehicle with registration year above 2016 can not be accurate. We also want to know the percetange of vehicles whose registration year was before 1900, because the car was invented on the earlies 1900's. 

In [71]:
sum(~autos["registration_year"].between(1900,2016))/autos.shape[0]

0.038793369710697

From that we can conclude that less than 4% of the vehicle's registration year was before "1900" therefore we will remove those rows. 

In [72]:
autos=autos[autos["registration_year"].between(1900,2016)]
print((autos["registration_year"]
                           .value_counts(normalize = True)
                           .sort_index(ascending = False)
                           .head(20)
                            ))

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
2006    0.057197
2005    0.062895
2004    0.057904
2003    0.057818
2002    0.053255
2001    0.056468
2000    0.067608
1999    0.062060
1998    0.050620
1997    0.041794
Name: registration_year, dtype: float64


It appears that most of the vehicles were first registered in the past 20 years.

# Exploring price by brand 

In [73]:
print(autos["brand"].unique().shape)
index=(autos["brand"]
      .value_counts(normalize = True)
      .sort_values(ascending = False).head(20))
print(index)

(40,)
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64


In [74]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [75]:
brand_mean_prices = {}
for brand in common_brands:
    brand_only = autos[autos["brand"]==brand]
    price_brand = brand_only["price"]
    brand_mean_prices[brand]=int(price_brand.mean())
brand_mean_prices    


{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

Of the top 5 brands, there is a distinct price gap 
* "Audi", "Mercedes Benz"  and "Bmw" are more expensive
* "Ford" and "Opel" are less expensive
* "Volkswagen"  is in between, that could be the reason that explains its popularity.

# Exploring mileage 

In [83]:
brand_mean_mileage={}
for brand in common_brands:
    brand_only=autos[ autos["brand"]==brand ]
    mileage_avg=brand_only["odometer_km"].mean()
    brand_mean_mileage[brand]=int(mileage_avg)    
brand_mean_mileage

{'audi': 129157,
 'bmw': 132572,
 'ford': 124266,
 'mercedes_benz': 130788,
 'opel': 129310,
 'volkswagen': 128707}

In [88]:
brand_mean_mileage_series=pd.Series(brand_mean_mileage)
brand_mean_prices_series=pd.Series(brand_mean_prices)

In [90]:
brand_mean_comparison=pd.DataFrame(
    brand_mean_prices_series,columns=['mean_price'])
brand_mean_comparison["mean_mileage"]=brand_mean_mileage_series
print(brand_mean_comparison)

               mean_price  mean_mileage
audi                 9336        129157
bmw                  8332        132572
ford                 3749        124266
mercedes_benz        8628        130788
opel                 2975        129310
volkswagen           5402        128707


The range of car mileages does not vary as much as the price do by brand. 